### PR TITLE
Rule expansion

### PR DESCRIPTION
**Pull request by knoxsp**
This expands the concept of rules to be much more expressive, for example to have rule types.

### CODE REVIEW

File: hydra_base/db/model.py [hunk 1]
```diff
@@ -96,6 +110,130 @@ class Inspect(object):
...
+                return False
+
+        return True
+
+    def check_user(self, user_id):
+        """
+            Check whether this user can read this dataset
+        """
+
+        if self.hidden == 'N':
```
[the remark this answers is not in the document]
Comment (knoxsp) [reply hydra_base/db/model.py]: I've removed the check_user function as it has been superceded by the check_read_permission function. The check_read_permission function is now called instead.

File: hydra_base/db/model.py [hunk 2]
```diff
@@ -96,6 +110,130 @@ class Inspect(object):
...
+        if _is_admin(user_id):
+            return True
+
+        if str(user_id) == str(self.created_by):
+            return True
+            
+        #used for datasets but could be extended to other objects, hence checking
+        #for the presence of hidden instead of adding a funciton to the subclass
+        if hasattr(self, 'hidden') and self.hidden == 'N':
+            return True
```
[the remark this answers is not in the document]
Comment (knoxsp) [reply hydra_base/db/model.py]: This has been resolved after a discussion and some code changes to put the 'hidden' check directly on the dataset object insead of the permission controlled superclass

File: hydra_base/db/model.py [hunk 3]
```diff
@@ -1480,22 +1681,25 @@ class Rule(Base, Inspect):
...
     name = Column(String(200), nullable=False)
-    description = Column(String(1000), nullable=False)
+    description = Column(String(1000), nullable=True)
+
+    format = Column(String(80), nullable=False, server_default='text')
 
-    cr_date = Column(TIMESTAMP(),  nullable=False, server_default=text(u'CURRENT_TIMESTAMP'))
     ref_key = Column(String(60),  nullable=False, index=True)
 
     value = Column(Text().with_variant(mysql.TEXT(4294967295), 'mysql'),  nullable=True)
```
[the remark this answers is not in the document]
Comment (knoxsp) [reply hydra_base/db/model.py]: I've done this across the board. Fingers crossed it won't break anthing!

**Comment by pmslavin** [comment]
The alembic upgrade script added in `ac09cd6` doesn't add the `created_at` column to the various `tRuleXXX` tables, causing an error when the audit mixin attempts to update these.

**Comment by pmslavin** [comment]
Seeing an error on Rule creation...

rule-create-error 00

`tRuleTypeLink` and `tRuleTypeDefinition` are both empty here.

**Pull request merged**